### PR TITLE
DAOS-5065 csum: Server Checksum Fetch Logic Rework

### PR DESCRIPTION
**Pull request by ryon-jensen**
This patch reworks the server side logic for retrieving
checksums on a fetch. It is more performant for larger
extents and has been simplified to be easier to read. Now
instead of iterating over each chunk and determining what
to do for each, the logic iterates biovs. If the biov
is a hole, then it skips it, otherwise it first considers
the prefix/suffix of a biov and if current index is in the
same chunk as the previx or suffix a new checksum is needed.
The logic also considers if there is a different non
hole biov in the same chunk. If a new checksum is needed
then the logic will continue iterating over biovs, using
the data to calculate the new checksum until the end of
the chunk or request is reached.

If a new checksum is not needed, then it will copy as much
of the biov as possible which would include all but
possibly the last chunk of the biov.

Also ...
- Updated checksum_timing utility
- For the srv checksum tests, changed how checksums are
  verified. Instead of checking specific csum results, now
  just looking to see if new checksums are created or copied
  from what's stored. This makes it easier to read (it's a
  string now) and verify the server checksum logic.
- Added several new server checksum tests
- Added more testing to checksum functional tests.
- Setup daos_checksum tests to test all checksum types
  with an env variable set. Don't want it to run that way in
  CI as it can be slow.
- Fixed bug where evtree function to update biov's with
  prefix/suffix and adjust csums wasn't taking record len
  into account. Was also double offsetting the
  checksum (evtree.c and vos_io.c)
- Updated data_integrity and readme documentation

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

File: src/common/checksum.c [hunk 1]
```diff
@@ -1419,7 +1419,9 @@ ci_insert(struct dcs_csum_info *dcb, int idx, uint8_t *csum_buf, size_t len)
 {
 	uint8_t *to_update;
 
-	D_ASSERT(idx < dcb->cs_nr);
+	D_ASSERTF(idx < dcb->cs_nr, "idx(%d) < dcb->cs_nr(%d)",
```
Comment (jolivier23) [inline src/common/checksum.c:1422]: I presume these can't be caused by user error?

File: src/common/tests/SConscript [hunk 2]
```diff
@@ -52,7 +52,7 @@ def scons():
                     LIBS=['daos_common', 'gurt', 'pmemobj', 'cmocka'])
     daos_build.test(tenv, 'other', 'other.c',
                     LIBS=['daos_common', 'gurt', 'cart'])
-    daos_build.test(tenv, 'common_test',
+    common_test = daos_build.test(tenv, 'common_test',
```
Comment (jolivier23) [inline src/common/tests/SConscript:55]: Not really your thing but we should probably just install all of these tests.  Though internal tests should probably go somewhere like install/TESTING/utest or something

File: src/common/tests/checksum_timing.c [hunk 3]
```diff
@@ -212,36 +270,72 @@ m_csum64_finish(struct daos_csummer *obj)
...
+		.cf_init = m_csum64_init,
+		.cf_destroy = m_csum64_destroy,
+		.cf_finish = m_csum64_finish,
+		.cf_get_size = m_csum64_get_size,
 		.name = "mcrc64"
 	};
 #endif
 
+static int
+gurt_64_update(struct daos_csummer *obj,
```
Comment (jolivier23) [inline src/common/tests/checksum_timing.c:284]: kind of a weird name

File: src/common/tests/checksum_timing.c [hunk 4]
```diff
@@ -212,36 +270,72 @@ m_csum64_finish(struct daos_csummer *obj)
...
+}
+
+struct csum_ft gurt_murmur64_algo = {
+	.cf_update = gurt_64_update,
+	.cf_csum_len = sizeof(uint64_t),
+	.cf_name = "murm64"
+};
+
+static int
+hash_string32_update(struct daos_csummer *obj,
```
Comment (jolivier23) [inline src/common/tests/checksum_timing.c:301]: same comment...

At least maybe make them consistent.  hash_string32* and hash_murmur*

File: src/common/tests/checksum_timing.c [hunk 5]
```diff
@@ -341,19 +450,23 @@ main(int argc, char *argv[])
 #ifdef MCHECKSUM_SUPPORT
```
Comment (jolivier23) [inline src/common/tests/checksum_timing.c:450]: do we still need mchecksum?

File: src/common/checksum.c (hunk 1 above)
Comment (ryon-jensen) [reply src/common/checksum.c:1422]: Shouldn't be able to. All calls to this are internal to the server.

File: src/common/tests/SConscript (hunk 2 above)
Comment (ryon-jensen) [reply src/common/tests/SConscript:55]: Good idea. I'll create a ticket for it. Rather not do it now so I can get this change in.

File: src/common/tests/checksum_timing.c (hunk 4 above)
Comment (ryon-jensen) [reply src/common/tests/checksum_timing.c:301]: changed to be more consistent and hopefully less weird :)

File: src/common/tests/checksum_timing.c (hunk 5 above)
Comment (ryon-jensen) [reply src/common/tests/checksum_timing.c:450]: It's interesting to me. This utility is a good way to compare different checksum libraries and implementations. For completion sake I left mchecksum in, but it's not really needed. I'll add a ticket to remove it in the future if you feel it adds too much noise.

**Comment by ryon-jensen** [comment]
Only failure is checkpatch and the issues don't seem to be significant enough to run through CI again. Is this something that can be landed or should I really address those white space issues?

**Comment by ashleypittman** [comment]
> Only failure is checkpatch and the issues don't seem to be significant enough to run through CI again. Is this something that can be landed or should I really address those white space issues?

The checkpatch warnings look trivial to fix and have been there in previous builds so there's been plenty of changes to fix them.

One of them looks wrong however so I'm going to try some tests on this PR to see what the problem is, if you get a checkpatch-style comment from my account shortly please ignore it.

File: src/object/tests/srv_checksum_tests.c [hunk 6]
```diff
@@ -814,25 +1390,56 @@ int teardown(void **state)
...
-	TA("05: Extent is larger than chunk", with_extent_larger_than_request),
-	TA("06: Fetch smaller than chunk", with_fetch_smaller_than_chunk),
-	TA("07: Partial extent/unaligned extent", more_partial_extent_tests),
-	TA("08: Fetch with larger records", test_larger_records),
-	TA("09: Fetch with larger records", test_larger_records2),
-	TA("10: Determine if need new checksum", need_new_checksum_tests),
+	TA("SRV_CSUM_ARRAY01: Whole extent requested",
+	   request_that_matches_single_extent),
+	TA("SRV_CSUM_ARRAY02: Whole extent requested, broken into multiple "
+	   "chunks", request_that_matches_single_extent_multiple_chunks),
```
Comment (ashleypittman) [inline src/object/tests/srv_checksum_tests.c:1401]: (style)  quoted string split across lines

File: src/object/tests/srv_checksum_tests.c [hunk 7]
```diff
@@ -814,25 +1390,56 @@ int teardown(void **state)
...
+	TA("SRV_CSUM_ARRAY03: Whole extent requested, broken into multiple "
+	   "chunks, not aligned to chunk",
```
Comment (ashleypittman) [inline src/object/tests/srv_checksum_tests.c:1403]: (style)  quoted string split across lines

File: src/tests/suite/daos_checksum.c [hunk 8]
```diff
@@ -1997,7 +2071,13 @@ static const struct CMUnitTest csum_tests[] = {
 	CSUM_TEST("DAOS_CSUM14: Many IODs", many_iovs_with_single_values),
 	CSUM_TEST("DAOS_CSUM15: Request non existent data",
 		  request_non_existent_data),
-
+	CSUM_TEST("DAOS_CSUM16: Unaligned hole at beginning",
+		  unaligned_hole_at_beginning),
+	CSUM_TEST("DAOS_CSUM17: Through some random testing found a rounding "
+		  "error", bug_rounding_error),
```
Comment (ashleypittman) [inline src/tests/suite/daos_checksum.c:2077]: (style)  quoted string split across lines

File: src/tests/suite/daos_checksum.c [hunk 9]
```diff
@@ -1997,7 +2071,13 @@ static const struct CMUnitTest csum_tests[] = {
...
+	CSUM_TEST("DAOS_CSUM18: request extent starts much later than the "
+		  "beginning of the stored extent",
```
Comment (ashleypittman) [inline src/tests/suite/daos_checksum.c:2079]: (style)  quoted string split across lines

File: src/tests/suite/daos_checksum.c [hunk 10]
```diff
@@ -2013,20 +2093,46 @@ static const struct CMUnitTest csum_tests[] = {
...
+		return 0;
+	}
+
+	if (sub_tests_size == 0) {
+		if (getenv("DAOS_CSUM_TEST_ALL_TYPE")) {
+			for (i = DAOS_PROP_CO_CSUM_OFF + 1;
+			     i <= DAOS_PROP_CO_CSUM_SHA512; i++) {
+				dts_csum_prop_type = i;
+				print_message("Running tests with csum_type: "
+					      "%d\n", i);
```
Comment (ashleypittman) [inline src/tests/suite/daos_checksum.c:2122]: (style)  quoted string split across lines

File: src/tests/suite/daos_test_common.c [hunk 11]
```diff
@@ -330,7 +330,9 @@ test_setup(void **state, unsigned int step, bool multi_rank,
 
 	/** Look at variables set by test arguments and setup container props */
 	if (dt_csum_type) {
-		printf("\n-------\nChecksum enabled in test!\n-------\n");
+		print_message("\n-------\n"
+			      "Checksum enabled in test!"
+			      "\n-------\n");
```
Comment (ashleypittman) [inline src/tests/suite/daos_test_common.c:335]: (style)  quoted string split across lines

**Comment by ryon-jensen** [comment]
assuming everyone is ok with the minor checkpatch fixes and since already passed CI requesting gatekeeper.

**Pull request merged**